### PR TITLE
Blog post about Mutiny and back pressure

### PR DESCRIPTION
**Pull request by cescoffier**
CC @jponge

### CODE REVIEW

File: _posts/2020-10-27-mutiny-back-pressure.adoc [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+---
+
+Reactive Programming offers an elegant, flexible, and powerful way to write asynchronous code.
+It supports failure handling, both sequential and parallel operation composition, and plenty of operators.
+Reactive programming promotes data streams as a primary construct, and your code is observing streams and reacting to signals. 
+
+.Origins of back-pressure (https://impurepics.com/posts/2018-05-26-backpressure-origins.html)   
+image::/assets/images/posts/mutiny-back-pressure/back-pressure-origin.png[800, 800]
+
+But, using data streams as primary constructs does not come without some issues.
```
Comment (jponge) [inline _posts/2020-10-27-mutiny-back-pressure.adoc:17]: But -> However

File: _posts/2020-10-27-mutiny-back-pressure.adoc [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+## The producer/consumer problem
+
+Let's imagine a fast producer and a slow consumer.
+The producer sends events too quickly for the consumer that can't keep up.
+That phenomenon is remarkably well illustrated in the following picture.
+
+.No Back-Pressure (https://impurepics.com/posts/2018-05-26-backpressure-origins.html)
+image::/assets/images/posts/mutiny-back-pressure/back-pressure-no.png[800, 800]
+
+But, let's see some code.
```
Comment (jponge) [inline _posts/2020-10-27-mutiny-back-pressure.adoc:29]: Maybe drop the `But,`

**Comment by jponge** [comment]
Interesting story, but it ended abruptly without an example of requesting items from a subscription, which is the whole point towards then end of the blog post 😉

**Comment by cescoffier** [comment]
Good idea! Let me add that. I don't like the end (seems like after 2h of writing, my inspiration level gets low)

**Comment by cescoffier** [comment]
I have added 2 code snippets at the end of the post explaining how you can control the flow yourself.

**Comment by jponge** [comment]
Much better now 👍

File: _posts/2020-10-27-mutiny-back-pressure.adoc [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+That control can use reducers or bends. 
+While very interesting, if you are a plumber, it's not clear how it can help us here? 
+
+We can see our streams as a flow of fluid, and the set of stages (operator or subscriber) forms a pipe. 
+We want to make the fluid flows as frictionless as possible without swirls and waves. 
+
+An interesting characteristic of fluid mechanics is how a downstream reduction of the throughput affects the upstream.  
+Essentially, that's what we need: a way for the downstream operators and subscribers to reduce the throughput, not only locally but also upstream. 
+
+Don't be mistaken; back-pressure is not something new in the IT world and it not limited to Reactive.
```
Comment (kenfinnigan) [inline _posts/2020-10-27-mutiny-back-pressure.adoc:151]: s/it/is ?

File: _posts/2020-10-27-mutiny-back-pressure.adoc [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+But, the consequence need to be understood: blocking the writer may not be without side-effects.
+
+There are other back-pressure implementations. 
+AMQP uses a credit-based flow control, where you can only send if you have enough credit.
+Eclipse Vert.x back-pressure is based on pause/resume, where a consumer can pause the upstream until it catches up and resume it once back on track. 
+
+### Introducing Reactive Streams
+
+Let's now focus on another back-pressure protocol: https://www.reactive-streams.org/[Reactive Streams].
+It defines an asynchronous and back-pressure protocol suiting perfectly to our fast producer/slow consumer problem.
```
Comment (kenfinnigan) [inline _posts/2020-10-27-mutiny-back-pressure.adoc:164]: Maybe reword "suiting perfectly to our fast", seems a little off.

Maybe "suited" is enough?

File: _posts/2020-10-27-mutiny-back-pressure.adoc [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    .onSubscribe().invoke(sub -> System.out.println("Received subscription: " + sub))
+    .onRequest().invoke(req -> System.out.println("Got a request: " + req))
+    .transform().byFilteringItemsWith(i -> i % 2 == 0)
+    .onItem().transform(i -> i * 100)
+    .subscribe().with(
+            i -> System.out.println("i: " + i)
+);
+---- 
+
+You can also going a step further and not only observing but also controlling the flow yourself:
```
Comment (kenfinnigan) [inline _posts/2020-10-27-mutiny-back-pressure.adoc:239]: s/going/go

File: _posts/2020-10-27-mutiny-back-pressure.adoc [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+The Reactive Streams protocol works well when you can control the pace of the upstream.
+But it's not always the case. 
+Streams representing physical entities are rarely controllable. 
+Imagine a stream emitting user's keystrokes.
+You cannot ask the users to be slow down. 
+That would be terrible user experience. 
+As we have seen above, time is also not something we can slow down, unfortunately...
+In this case, the `onOverflow()` group lets you decide the mitigation, such as using buffers or dropping items. 
+
+It's imperative to avoid overwhelming downstream subscribers.
```
Comment (kenfinnigan) [inline _posts/2020-10-27-mutiny-back-pressure.adoc:312]: I'd maybe change "imperative" to "critical", or something similar, to prevent any possible confusion around imperative vs reactive code.

**Review by kenfinnigan — CHANGES_REQUESTED** [review]

File: _posts/2020-10-27-mutiny-back-pressure.adoc (hunk 6 above)
Comment (cescoffier) [reply _posts/2020-10-27-mutiny-back-pressure.adoc:312]: good point!

**Comment by cescoffier** [comment]
@kenfinnigan All done! Thanks!

**Review by kenfinnigan — APPROVED** [review]
